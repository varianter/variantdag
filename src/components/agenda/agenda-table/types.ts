import { FeatureSlot } from './feature-slot/FeatureSlot';
import { RowSlot } from './row-slot/RowSlot';

/**
 * Duration created from hours and minutes used for relative time calculation.
 *
 * Example: 10:30 => 10*60 minutes + 30 minutes = 630 minutes
 */
export type DurationMinutes = number;

export type Time = {
  hour: number;
  minutes: number;
};

/**
 * Location of a feature, such as a presentation or meeting:
 *
 * Examples: "Oktetten", "Amfiet"
 */
export type VenueName = string;

/**
 * A time slot in the program
 */
export interface Slot {
  from: Time;
  to: Time;
}

/**
 * The entire program of a conference. The variables 'from' and 'to' describes
 * the start and end times of the program, and is necessary for style calculations.
 *
 * The base interfaces FeatureSlot and RowSlot contains the properties required by AgendaTable.
 * However, in order to allow writing render functions for Row and Feature slots, the base
 * interfaces may be extended and used instead.
 */
export type Program<
  T extends FeatureSlot = FeatureSlot,
  K extends RowSlot = RowSlot,
> = {
  from: Time;
  to: Time;
  features: T[];
  rows: K[];
};

/**
 * Any function that implements the render functions, can be used to control how
 * rows, features and headers are displayed in the AgendaTable. Extend the RowSlot and FeatureSlot
 * interfaces to allow rendering with additional properties.
 */
export type RowRenderer<T extends RowSlot> = (row: T) => JSX.Element;

export type FeatureRenderer<T extends FeatureSlot> = (
  feature: T,
) => JSX.Element;

export type HeaderRenderer = (venueName: VenueName) => JSX.Element;
