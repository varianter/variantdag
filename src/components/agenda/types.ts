/**
 * Duration created from hours and minutes used for relative time calculation.
 *
 * Example: 10:30 => 10*60 minutes + 30 minutes = 630 minutes
 */
export type DurationMinutes = number;

export type Time = {
  hour: number;
  minutes: number;
};

/**
 * Location of a feature, such as a presentation or meeting:
 *
 * Examples: "Oktetten", "Amfiet"
 */
export type Venue = string; // e.g

/**
 * A time slot in the program
 */
export interface Slot {
  from: Time;
  to: Time;
}

/**
 * A presentation, meeting or similar
 */
export interface Feature extends Slot {
  title: string;
  venue: Venue;
}

/**
 * The entire program of a conference. The variables 'from' and 'to' describes
 * the start and end times of the program, and is necessary for style calculations.
 */
export type Program = {
  from: Time;
  to: Time;
  features: Feature[];
};
