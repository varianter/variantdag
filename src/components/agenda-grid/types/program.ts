import { Feature } from '../feature/Feature';
import { Time } from './time';

/**
 * Agenda program containg Features (e.g. talks, meetings, workshops)
 */
export type Program = {
  from: Time;
  to: Time;
  features: Feature[];
};
