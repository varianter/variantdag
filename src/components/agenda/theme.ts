import { colorPairs } from '@variant/profile/lib/colors';
import { CSSProperties } from 'react';

export type Color = string;

/**
 * Describes how a Row or Feature should look
 */
export interface ThemeType {
  text: Color;
  background: Color;
}

export const Theme = {
  FELLES: {
    text: colorPairs.primary.default.text,
    background: colorPairs.primary.default.bg,
  } as ThemeType,
  LÃ†RE: {
    text: colorPairs.secondary1.default.text,
    background: colorPairs.secondary1.default.bg,
  } as ThemeType,
  SKUDD: {
    text: colorPairs.secondary2.default.text,
    background: colorPairs.secondary2.default.bg,
  } as ThemeType,
  ANNET: {
    text: colorPairs.secondary4.default.text,
    background: colorPairs.secondary4.default.bg,
  } as ThemeType,
  PAUSE: {
    text: colorPairs.secondary3.default.text,
    background: colorPairs.secondary3.default.bg,
  } as ThemeType,
};

export type ThemeName = keyof typeof Theme;

export const toStyle = (theme: ThemeType): CSSProperties => ({
  color: theme.text,
  backgroundColor: theme.background,
});
