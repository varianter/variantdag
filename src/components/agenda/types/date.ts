export type Month =
  | '01'
  | '02'
  | '03'
  | '04'
  | '05'
  | '06'
  | '07'
  | '08'
  | '09'
  | '10'
  | 11
  | 12;

export type FirstDayDigit = 0 | 1 | 2 | 3;
export type SecondDayDigit = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

export type Date =
  `20${number}${number}-${Month}-${FirstDayDigit}${SecondDayDigit}`;
