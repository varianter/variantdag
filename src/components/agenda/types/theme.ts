import { colorPairs } from '@variant/profile/lib/colors';
import { CSSProperties } from 'react';

export type Color = string;
export interface ThemeType {
  text: Color;
  background: Color;
}

/**
 * Contains color pairs for text and background suitable for agenda events of
 * different types.
 * */
export const Theme = {
  FELLES: {
    text: colorPairs.primary.default.text,
    background: colorPairs.primary.default.bg,
  } as ThemeType,
  LÃ†RE: {
    text: colorPairs.secondary1.default.text,
    background: colorPairs.secondary1.default.bg,
  } as ThemeType,
  SKUDD: {
    text: colorPairs.secondary2.default.text,
    background: colorPairs.secondary2.default.bg,
  } as ThemeType,
  ANNET: {
    text: colorPairs.secondary4.default.text,
    background: colorPairs.secondary4.default.bg,
  } as ThemeType,
  PAUSE: {
    text: colorPairs.secondary3.default.text,
    background: colorPairs.secondary3.default.bg,
  } as ThemeType,
};

export type ThemeName = keyof typeof Theme;

export const toStyle = (theme: ThemeType): CSSProperties => ({
  color: theme.text,
  backgroundColor: theme.background,
});
