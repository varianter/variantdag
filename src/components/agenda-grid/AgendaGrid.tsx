import styles from './AgendaGrid.module.css';
import { Feature } from './feature/Feature';
import { Program } from './types/program';
import { getUniqueVenues } from './utils/featureUtils';
import {
  calculateFeaturePosition,
  calculateGridLayout,
} from './utils/gridLayout';

type Props = {
  program: Program;
};

export const AgendaGrid = ({ program }: Props) => {
  const uniqueVenues = getUniqueVenues(program.features);

  const features = program.features.map((feature) => (
    <Feature
      feature={feature}
      style={calculateFeaturePosition(feature.from, feature.to, feature.venue)}
      key={`${feature.venue}-${feature.from}-${feature.to}`}
    />
  ));

  return (
    <section
      style={calculateGridLayout(program.from, program.to, uniqueVenues)}
      className={styles.grid}
    >
      {features}
    </section>
  );
};
