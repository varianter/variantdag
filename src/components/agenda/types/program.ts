import { Feature } from '../feature/Feature';
import { Time } from './time';
import { Row } from '../row/Row';
import { ThemeType } from './theme';
import { Date } from './date';

/**
 * Agenda program containg Features (e.g. talks, meetings, workshops)
 */
export type Program = {
  date: Date;
  from: Time;
  to: Time;
  events: (Event | VenueEvent)[];
};

export type Event = {
  title: string;
  from: Time;
  to: Time;
  theme: ThemeType;
  description?: string;
};

export type VenueEvent = Event & {
  venue: string;
};
