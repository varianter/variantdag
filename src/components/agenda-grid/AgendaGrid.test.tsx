import { AgendaGrid } from './AgendaGrid';
import renderer from 'react-test-renderer';
import { Program } from './types/program';
import { Theme } from './types/theme';

const program: Program = {
  from: '07:45',
  to: '11:30',
  features: [
    {
      title: 'Kaffe og mingling',
      from: '07:45',
      to: '08:00',
      venue: 'Amfiet',
      theme: Theme.PAUSE,
    },
    {
      title: 'Velkommen',
      from: '08:15',
      to: '09:00',
      venue: 'Amfiet',
      theme: Theme.FELLES,
    },
    {
      title: 'Hvordan gjør kunde fornøyd når man sier ting de ikke vil høre?',
      from: '09:15',
      to: '11:30',
      venue: 'Amfiet',
      theme: Theme.SKUDD,
    },
    {
      title:
        'Workshop om mye spennende opplegg som skal beskrives med veldig, veldig, veldig mange ord.',
      from: '09:15',
      to: '10:00',
      venue: 'Torsken',
      theme: Theme.LÆRE,
    },
    {
      title: 'Egenarbeid',
      from: '10:00',
      to: '11:30',
      venue: 'Torsken',
      theme: Theme.ANNET,
    },
  ],
};

describe('AgendaGrid', () => {
  it('renders a populated agenda', () => {
    const agenda = renderer.create(<AgendaGrid program={program} />);

    expect(agenda).toMatchSnapshot();
  });
});
