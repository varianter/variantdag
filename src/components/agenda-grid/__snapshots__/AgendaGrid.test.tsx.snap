// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgendaGrid renders a populated agenda 1`] = `
<section
  className="grid"
  style={
    Object {
      "gridTemplateColumns": " [Amfiet] 1fr [Torsken] 1fr",
      "gridTemplateRows": " [kl-0745] 1fr [kl-0800] 1fr [kl-0815] 1fr [kl-0830] 1fr [kl-0845] 1fr [kl-0900] 1fr [kl-0915] 1fr [kl-0930] 1fr [kl-0945] 1fr [kl-1000] 1fr [kl-1015] 1fr [kl-1030] 1fr [kl-1045] 1fr [kl-1100] 1fr [kl-1115] 1fr",
    }
  }
>
  <div
    className="feature"
    style={
      Object {
        "backgroundColor": "#FFC4BC",
        "color": "#333333",
        "gridColumn": "Amfiet",
        "gridRowEnd": "kl-0800",
        "gridRowStart": "kl-0745",
      }
    }
  >
    <div>
      07:45
       - 
      08:00
    </div>
    <div
      className="title"
    >
       
      Kaffe og mingling
    </div>
  </div>
  <div
    className="feature"
    style={
      Object {
        "backgroundColor": "#E61A6B",
        "color": "#FFFFFF",
        "gridColumn": "Amfiet",
        "gridRowEnd": "kl-0900",
        "gridRowStart": "kl-0815",
      }
    }
  >
    <div>
      08:15
       - 
      09:00
    </div>
    <div
      className="title"
    >
       
      Velkommen
    </div>
  </div>
  <div
    className="feature"
    style={
      Object {
        "backgroundColor": "#03DAC6",
        "color": "#333333",
        "gridColumn": "Amfiet",
        "gridRowEnd": "kl-1130",
        "gridRowStart": "kl-0915",
      }
    }
  >
    <div>
      09:15
       - 
      11:30
    </div>
    <div
      className="title"
    >
       
      Hvordan gjør kunde fornøyd når man sier ting de ikke vil høre?
    </div>
  </div>
  <div
    className="feature"
    style={
      Object {
        "backgroundColor": "#423D89",
        "color": "#FFFFFF",
        "gridColumn": "Torsken",
        "gridRowEnd": "kl-1000",
        "gridRowStart": "kl-0915",
      }
    }
  >
    <div>
      09:15
       - 
      10:00
    </div>
    <div
      className="title"
    >
       
      Workshop om mye spennende opplegg som skal beskrives med veldig, veldig, veldig mange ord.
    </div>
  </div>
  <div
    className="feature"
    style={
      Object {
        "backgroundColor": "#EDE8D7",
        "color": "#333333",
        "gridColumn": "Torsken",
        "gridRowEnd": "kl-1130",
        "gridRowStart": "kl-1000",
      }
    }
  >
    <div>
      10:00
       - 
      11:30
    </div>
    <div
      className="title"
    >
       
      Egenarbeid
    </div>
  </div>
</section>
`;
